azure-iot-device/azure/iot/device/iothub/transport/mqtt/mqtt_topic.py

```diff
@@ -46,6 +46,24 @@ def get_input_topic_for_subscribe(device_id, module_id):
     return _get_topic_base(device_id, module_id) + "/inputs/#"
 
 
+def get_method_topic_for_publish(request_id, status):
+    """
+    :return: The topic for publishing method responses. It is of the format
+    "$iothub/methods/res/<status>/?$rid=<requestId>
+    """
+    return "$iothub/methods/res/{status}/?$rid={rid}".format(
+        status=urllib.parse.quote(status), rid=urllib.parse.quote(request_id)
```

Comment:
>status=urllib.parse.quote(status), rid=urllib.parse.quote(request_id) [](start = 8, length = 69)

you might be right about uri-encoding here, but this makes me wonder.  status is an integer, and request_id needs to be mirrored from the request.  encoding status is, at worst, an unnecessary function call, but encoding request_id _might_ be broken if urllib.parse.quote isn't an exact inverse of unquote_plus (which is used to extract the rid from the request).  (this assumes request id's can be arbitrary, but they're defined by the iothub service, so it may never _need_ encoding/decoding)

Reply:
I'll switch them to both to use the quote_plus variant. 

I agree that the field is probably just fine on it's own (rids I think are just integers) and there's probably no need to ever do a "quote" on it, but since it's being decoded with unquote_plus due to being in a function that has to be more generic, it makes sense to keep it all symmetrical.

We could **probably** drop the quote on status, but I also just generally think url encoding/decoding everything going into a URI is generally good practice - not just from a defensive coding perspective, but also because it makes it easier to notice when url encoding is missing, and correct it.

i.e. If I'm scanning a file, and I notice something in a uri isn't being encoded/decoded, I know it should be fixed - I don't have to think about "should this particular field I may not know about be encoded/decoded?". This prevents any bugs with things that DO need to be encoded/decoded elsewhere from slipping through.

Open to doing it differently though, let me know.

---
In reply to: [282744164](https://github.com/Azure/azure-iot-sdk-python-preview/pull/86#discussion_r282744164) [](ancestors = 282744164)